src/Runner/Runner.php

```diff
@@ -89,26 +123,250 @@ public function __construct(
+
+                $process->request(['action' => ParallelAction::RUNNER_REQUEST_ANALYSIS, 'files' => $fileChunk]);
+            });
+        });
+
+        $processesToSpawn = min(
+            $this->parallelConfig->getMaxProcesses(),
+            (int) ceil($this->fileCount / $this->parallelConfig->getFilesPerProcess())
         );
+        $processFactory = new ProcessFactory($this->input);
+
+        for ($i = 0; $i < $processesToSpawn; ++$i) {
+            $identifier = ProcessIdentifier::create();
+            $process = $processFactory->create(
+                $streamSelectLoop,
+                new RunnerConfig(
+                    $this->isDryRun,
+                    $this->stopOnViolation,
+                    $this->parallelConfig,
+                    $this->configFile
+                ),
+                $identifier,
+                $serverPort,
+            );
+            $processPool->addProcess($identifier, $process);
+            $process->start(
+                // [REACT] Handle workers' responses (multiple actions possible)
+                function (array $workerResponse) use ($processPool, $process, $identifier, $getFileChunk, &$changed): void {
+                    // File analysis result (we want close-to-realtime progress with frequent cache savings)
+                    if (ParallelAction::WORKER_RESULT === $workerResponse['action']) {
+                        $fileAbsolutePath = $workerResponse['file'];
+                        $fileRelativePath = $this->directory->getRelativePathTo($fileAbsolutePath);
+
+                        // Dispatch an event for each file processed and dispatch its status (required for progress output)
+                        $this->dispatchEvent(
+                            FixerFileProcessedEvent::NAME,
+                            new FixerFileProcessedEvent($workerResponse['status'])
+                        );
+
+                        if (isset($workerResponse['fileHash'])) {
+                            $this->cacheManager->setFileHash($fileRelativePath, $workerResponse['fileHash']);
+                        }
+
+                        // Pass-back information about applied changes (only if there are any)
+                        if (isset($workerResponse['fixInfo'])) {
+                            $changed[$fileRelativePath] = $workerResponse['fixInfo'];
+
+                            if ($this->stopOnViolation) {
+                                $processPool->endAll();
+
+                                return;
+                            }
+                        }
+
+                        foreach ($workerResponse['errors'] ?? [] as $error) {
+                            $this->errorsManager->report(new Error(
+                                $error['type'],
+                                $error['filePath'],
+                                null !== $error['source']
+                                    ? ParallelisationException::forWorkerError($error['source'])
```

Comment:
The errors we are reporting in `WORKER_RESULT` action, [here](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7777/files#diff-451d731f59fc56df00f66e1873fb0800438813b1f87bba0ed062dc801ee81a0cR176), are not WorkerErrors, right ?

Reply:
I knew you'll point it out 😅. Yeah, these are process errors (exceptions caught during fixing, reported to `ErrorsManager::report()`), so I'll improve the naming accordingly.

Reply:
I try to review dilligently

Reply:
@keradus in terms of these errors, it's a little bit complicated. Main process spawns child processes, each worker runs `Runner` which applies rules and reports errors e.g. for linting errors, these errors are collected from `ErrorsManager` and reported through TCP socket to main process, which then creates `ParallelisationException` from raw data. This is basically dirty workaround just to display errors in the output:

<img width="868" alt="image" src="https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/assets/600668/1a9afabb-2299-4955-b910-77e09736344d">

BUT it's not `ParallelisationException` per se (as it's a caught exception during fixing process), and provided stacktrace is not a real one. I am not sure if we can recreate actual exception even if we report FQN of it from worker. Even if it's possible, then we won't recreate original stacktrace, because it's not something we can mimic when we create an exception. We can display the type of error though:

<img width="868" alt="image" src="https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/assets/600668/deb827ea-8739-483a-93fb-f3aa1f94addc">

For this particular reason I just dump original stacktrace as string in `WorkerError` (for unhandled exceptions), so [I can display it in the output with real trace](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7777#issuecomment-1972320854). In case of `Error` probably we can't just change it now 🤔.

Would you be OK with changing this to `ParallelisationException::forHandledErrorReportedByWorker()` and displaying like shown above, at least for now? We can improve it later IMHO, if possible.

Reply:
Please look [here](#discussion_r1595309165) for better context, as for providing process errors from workers to main process we have exact same possibilities.

Reply:
https://github.com/Wirone/PHP-CS-Fixer/pull/11

Reply:
Error handling was reworked, we can close this thread.